File: airs_experiment/R/run_11_typology.R
#!/usr/bin/env Rscript
# ==============================================================================
# AIRS Phase 11: Practitioner Typology Classification
# R port of run_11_typology.py
#
# K-means cluster analysis on AIRS-16 constructs.
# k=3 optimal, k=4 comparison, predictive validation, classification function.
# Outputs: results/phase11_practitioner_typology_results.json
# ==============================================================================

library(jsonlite)
library(cluster)

cat(strrep("=", 90), "\n")
cat("PHASE 11: PRACTITIONER TYPOLOGY CLASSIFICATION\n")
cat(strrep("=", 90), "\n")

dir.create("results", showWarnings = FALSE)

# ── Configuration ─────────────────────────────────────────────────────────────
AIRS16_CONSTRUCTS <- c("PE", "EE", "SI", "FC", "HM", "PV", "HB", "TR")
OUTCOME <- "BI"
SELECTED_K <- 3
K_ALT <- 4
RANDOM_STATE <- 42

# ── Load Data ─────────────────────────────────────────────────────────────────
df <- read.csv("data/AIRS_clean.csv", stringsAsFactors = FALSE)
cat(sprintf("\nLoaded: %d rows x %d columns\n", nrow(df), ncol(df)))

item_map <- list(
    PE = c("PE1", "PE2"), EE = c("EE1", "EE2"), SI = c("SI1", "SI2"),
    FC = c("FC1", "FC2"), HM = c("HM1", "HM2"), PV = c("PV1", "PV2"),
    HB = c("HB1", "HB2"), TR = c("TR1", "TR2")
)
for (cn in names(item_map)) {
    df[[cn]] <- rowMeans(df[, item_map[[cn]], drop = FALSE], na.rm = TRUE)
}
df$BI <- rowMeans(df[, c("BI1", "BI2", "BI3", "BI4")], na.rm = TRUE)

cat(sprintf("\nConstruct scores computed (N = %d)\n", nrow(df)))
print(summary(df[, c(AIRS16_CONSTRUCTS, "BI")]))

# Correlation with BI
cat("\nCorrelation with BI:\n")
for (cn in AIRS16_CONSTRUCTS) {
    r <- cor(df[[cn]], df$BI, use = "complete.obs")
    cat(sprintf("  %s: r = %.3f\n", cn, r))
}

# ── Prepare Clustering ───────────────────────────────────────────────────────
X <- df[complete.cases(df[, AIRS16_CONSTRUCTS]), AIRS16_CONSTRUCTS]
X_scaled <- scale(X)
pop_means <- colMeans(df[, AIRS16_CONSTRUCTS], na.rm = TRUE)
pop_stds <- apply(df[, AIRS16_CONSTRUCTS], 2, sd, na.rm = TRUE)

# ── Silhouette Analysis (k=2 to 8) ──────────────────────────────────────────
cat(sprintf("\n%s\n", strrep("=", 90)))
cat("SILHOUETTE ANALYSIS\n")
cat(strrep("=", 90), "\n")

sil_scores <- numeric()
for (k in 2:8) {
    set.seed(RANDOM_STATE)
    km <- kmeans(X_scaled, centers = k, nstart = 10)
    sil <- mean(silhouette(km$cluster, dist(X_scaled))[, "sil_width"])
    sil_scores[k] <- sil
    marker <- ifelse(k == SELECTED_K, " <-", "")
    cat(sprintf("  k=%d: silhouette = %.3f%s\n", k, sil, marker))
}

# ── k=3 Final Clustering ────────────────────────────────────────────────────
cat(sprintf("\n%s\n", strrep("=", 90)))
cat(sprintf("FINAL CLUSTERING (k=%d)\n", SELECTED_K))
cat(strrep("=", 90), "\n")

set.seed(RANDOM_STATE)
km_final <- kmeans(X_scaled, centers = SELECTED_K, nstart = 10)
df$Cluster <- NA
df$Cluster[as.integer(rownames(X))] <- km_final$cluster
final_silhouette <- mean(silhouette(km_final$cluster, dist(X_scaled))[, "sil_width"])
cat(sprintf("  Silhouette score: %.3f\n", final_silhouette))

# Cluster sizes
sizes <- table(df$Cluster)
for (c_id in names(sizes)) {
    cat(sprintf("  Cluster %s: n = %d (%.1f%%)\n", c_id, sizes[c_id], sizes[c_id] / nrow(df) * 100))
}

# Cluster profiles (z-scores)
cluster_means_raw <- aggregate(. ~ Cluster, data = df[!is.na(df$Cluster), c("Cluster", AIRS16_CONSTRUCTS)], FUN = mean)
cluster_means_z <- cluster_means_raw
for (cn in AIRS16_CONSTRUCTS) {
    cluster_means_z[[cn]] <- (cluster_means_raw[[cn]] - pop_means[cn]) / pop_stds[cn]
}

cat("\n  Z-scored cluster profiles:\n")
print(round(cluster_means_z, 3))

# Cluster naming
CLUSTER_NAMES <- character(SELECTED_K)
for (cid in 1:SELECTED_K) {
    profile <- cluster_means_z[cluster_means_z$Cluster == cid, AIRS16_CONSTRUCTS]
    mean_z <- mean(as.numeric(profile))
    if (mean_z > 0.3) {
        CLUSTER_NAMES[cid] <- "AI Enthusiasts"
    } else if (mean_z < -0.3) {
        CLUSTER_NAMES[cid] <- "AI Skeptics"
    } else {
        CLUSTER_NAMES[cid] <- "Moderate Users"
    }
}
df$Typology <- CLUSTER_NAMES[df$Cluster]

cat("\n  Cluster names:\n")
for (cid in 1:SELECTED_K) {
    n <- sizes[as.character(cid)]
    bi_mean <- mean(df$BI[df$Cluster == cid], na.rm = TRUE)
    cat(sprintf("  %d: %s (n=%d, BI mean=%.2f)\n", cid, CLUSTER_NAMES[cid], n, bi_mean))
}

# ── k=4 Comparison ──────────────────────────────────────────────────────────
cat(sprintf("\n%s\n", strrep("=", 90)))
cat(sprintf("k=%d ALTERNATIVE COMPARISON\n", K_ALT))
cat(strrep("=", 90), "\n")

set.seed(RANDOM_STATE)
km_k4 <- kmeans(X_scaled, centers = K_ALT, nstart = 10)
sil_k4 <- mean(silhouette(km_k4$cluster, dist(X_scaled))[, "sil_width"])
k4_sizes <- table(km_k4$cluster)
cat(sprintf("  k=4 silhouette: %.3f (k=3: %.3f)\n", sil_k4, final_silhouette))
for (c_id in names(k4_sizes)) {
    cat(sprintf("  Cluster %s: n = %d (%.1f%%)\n", c_id, k4_sizes[c_id], k4_sizes[c_id] / nrow(X) * 100))
}

# k=4 z-scored profiles
df_k4 <- df[!is.na(df$Cluster), ]
df_k4$K4_Cluster <- NA
df_k4$K4_Cluster[as.integer(rownames(X))] <- km_k4$cluster
k4_means <- aggregate(. ~ K4_Cluster, data = df_k4[!is.na(df_k4$K4_Cluster), c("K4_Cluster", AIRS16_CONSTRUCTS)], FUN = mean)
k4_means_z <- k4_means
for (cn in AIRS16_CONSTRUCTS) {
    k4_means_z[[cn]] <- (k4_means[[cn]] - pop_means[cn]) / pop_stds[cn]
}
cat("\n  k=4 z-scored profiles:\n")
print(round(k4_means_z, 3))

k4_defensible <- sil_k4 >= 0.25 && min(k4_sizes) >= 50
cat(sprintf(
    "\n  k=4 defensible: %s (sil>=0.25=%s, min_n>=50=%s)\n",
    k4_defensible, sil_k4 >= 0.25, min(k4_sizes) >= 50
))

# ── Predictive Validation ───────────────────────────────────────────────────
cat(sprintf("\n%s\n", strrep("=", 90)))
cat("PREDICTIVE VALIDATION\n")
cat(strrep("=", 90), "\n")

# ANOVA: BI by cluster
groups <- lapply(1:SELECTED_K, function(c) df$BI[df$Cluster == c & !is.na(df$Cluster)])
aov_res <- oneway.test(BI ~ factor(Cluster), data = df[!is.na(df$Cluster), ])
grand_mean <- mean(df$BI, na.rm = TRUE)
ss_between <- sum(sapply(groups, function(g) length(na.omit(g)) * (mean(g, na.rm = TRUE) - grand_mean)^2))
ss_total <- sum((df$BI[!is.na(df$Cluster)] - grand_mean)^2, na.rm = TRUE)
eta_squared <- ss_between / ss_total

effect_interp <- ifelse(eta_squared > 0.14, "large", ifelse(eta_squared > 0.06, "medium", "small"))
cat(sprintf("  ANOVA: F = %.2f, p = %.6f\n", aov_res$statistic, aov_res$p.value))
cat(sprintf("  eta2 = %.3f (%s)\n", eta_squared, effect_interp))

# Regression: AIRS-16 -> BI
X_reg <- df[complete.cases(df[, c(AIRS16_CONSTRUCTS, "BI")]), AIRS16_CONSTRUCTS]
y_reg <- df[rownames(X_reg), "BI"]
reg_fit <- lm(y_reg ~ ., data = X_reg)
r2 <- summary(reg_fit)$r.squared

# 5-fold CV R2
set.seed(RANDOM_STATE)
n <- nrow(X_reg)
folds <- sample(rep(1:5, length.out = n))
cv_r2 <- numeric(5)
for (fold in 1:5) {
    train <- folds != fold
    test <- folds == fold
    fit_cv <- lm(y_reg[train] ~ ., data = X_reg[train, ])
    preds <- predict(fit_cv, newdata = X_reg[test, ])
    ss_res <- sum((y_reg[test] - preds)^2)
    ss_tot <- sum((y_reg[test] - mean(y_reg[test]))^2)
    cv_r2[fold] <- 1 - ss_res / ss_tot
}
cat(sprintf("  Regression R2 = %.3f\n", r2))
cat(sprintf("  5-fold CV R2 = %.3f (+/-%.3f)\n", mean(cv_r2), sd(cv_r2)))

# Cluster-based R2
bi_cluster_mean <- tapply(df$BI[!is.na(df$Cluster)], df$Cluster[!is.na(df$Cluster)], mean)
df$BI_pred_cluster <- bi_cluster_mean[as.character(df$Cluster)]
ss_res_cluster <- sum((df$BI[!is.na(df$Cluster)] - df$BI_pred_cluster[!is.na(df$Cluster)])^2, na.rm = TRUE)
r2_cluster <- 1 - ss_res_cluster / ss_total
cat(sprintf("  Cluster-based R2 = %.3f\n", r2_cluster))

# ── Classification Function Validation ───────────────────────────────────────
# Store centroids for classification
CENTROIDS <- list()
for (cid in 1:SELECTED_K) {
    name <- CLUSTER_NAMES[cid]
    z_scores <- as.list(cluster_means_z[cluster_means_z$Cluster == cid, AIRS16_CONSTRUCTS])
    raw_means <- as.list(cluster_means_raw[cluster_means_raw$Cluster == cid, AIRS16_CONSTRUCTS])
    mask <- !is.na(df$Cluster) & df$Cluster == cid
    CENTROIDS[[name]] <- list(
        cluster_id = cid, z_scores = z_scores, raw_means = raw_means,
        expected_BI = mean(df$BI[mask], na.rm = TRUE),
        n = sum(mask), pct = sum(mask) / nrow(df) * 100
    )
}

classify_typology <- function(scores) {
    z <- (scores - pop_means) / pop_stds
    distances <- sapply(CENTROIDS, function(cd) {
        centroid_z <- unlist(cd$z_scores)
        sqrt(sum((z - centroid_z)^2))
    })
    nearest <- names(which.min(distances))
    list(type = nearest, confidence = 1 / (1 + min(distances)))
}

# Validate reclassification
correct <- 0
valid_rows <- which(!is.na(df$Cluster))
for (idx in valid_rows) {
    scores <- as.numeric(df[idx, AIRS16_CONSTRUCTS])
    names(scores) <- AIRS16_CONSTRUCTS
    pred <- classify_typology(scores)
    if (pred$type == df$Typology[idx]) correct <- correct + 1
}
agreement <- correct / length(valid_rows)
cat(sprintf("\n  Classification validation: %.1f%% agreement\n", agreement * 100))

# ── Success Criteria ─────────────────────────────────────────────────────────
cat(sprintf("\n%s\n", strrep("=", 90)))
cat("SUCCESS CRITERIA\n")
cat(strrep("=", 90), "\n")
cat(sprintf("  Silhouette > 0.25: %s (%.3f)\n", ifelse(final_silhouette > 0.25, "PASS", "FAIL"), final_silhouette))
cat(sprintf("  BI eta2 > 0.14: %s (%.3f)\n", ifelse(eta_squared > 0.14, "PASS", "FAIL"), eta_squared))
cat(sprintf("  Regression R2 > 0.40: %s (%.3f)\n", ifelse(r2 > 0.40, "PASS", "FAIL"), r2))

# ── Export ───────────────────────────────────────────────────────────────────
cat(sprintf("\n%s\n", strrep("=", 90)))
cat("EXPORTING RESULTS\n")
cat(strrep("=", 90), "\n")

results <- list(
    metadata = list(
        notebook = "11_Practitioner_Typology",
        date = as.character(Sys.time()),
        n_samples = nrow(df),
        constructs_used = AIRS16_CONSTRUCTS,
        outcome_variable = OUTCOME
    ),
    clustering = list(
        n_clusters = SELECTED_K,
        silhouette_score = final_silhouette,
        cluster_sizes = as.list(sizes)
    ),
    k4_alternative = list(
        n_clusters = K_ALT,
        silhouette_score = sil_k4,
        cluster_sizes = as.list(k4_sizes),
        defensible = k4_defensible
    ),
    validation = list(
        bi_anova_f = as.numeric(aov_res$statistic),
        bi_anova_p = aov_res$p.value,
        bi_eta_squared = eta_squared,
        regression_r2 = r2,
        regression_cv_r2 = mean(cv_r2),
        cluster_r2 = r2_cluster
    ),
    population_parameters = list(means = as.list(pop_means), stds = as.list(pop_stds)),
    centroids = CENTROIDS,
    cluster_names = as.list(CLUSTER_NAMES)
)

write_json(results, "results/phase11_practitioner_typology_results.json",
    pretty = TRUE, auto_unbox = TRUE
)
cat("Saved results/phase11_practitioner_typology_results.json\n")

cat(sprintf("\n%s\n", strrep("=", 90)))
cat("PHASE 11: PRACTITIONER TYPOLOGY COMPLETE\n")
cat(strrep("=", 90), "\n")
for (name in names(CENTROIDS)) {
    cd <- CENTROIDS[[name]]
    cat(sprintf("  %s: n=%d (%.1f%%), BI=%.2f\n", name, cd$n, cd$pct, cd$expected_BI))
}

File: airs_experiment/python/run_11_typology.py
#!/usr/bin/env python3
"""
AIRS Phase 11: Practitioner Typology Classification
Refactored from 11_Practitioner_Typology.ipynb

K-means cluster analysis for user typology based on AIRS-16 constructs.
Includes k=3 optimal solution, k=4 comparison, predictive validation,
and classification function for field deployment.
Outputs: results/phase11_practitioner_typology_results.json
"""

import json
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
from scipy import stats as scipy_stats
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import silhouette_score
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression

warnings.filterwarnings('ignore')

# ═══════════════════════════════════════════════════════════════════════════════
# CONFIGURATION
# ═══════════════════════════════════════════════════════════════════════════════

AIRS16_CONSTRUCTS = ['PE', 'EE', 'SI', 'FC', 'HM', 'PV', 'HB', 'TR']
OUTCOME = 'BI'
SELECTED_K = 3
K_ALT = 4
RANDOM_STATE = 42


def main():
    print("=" * 90)
    print("PHASE 11: PRACTITIONER TYPOLOGY CLASSIFICATION")
    print("=" * 90)

    # ── Load Data ─────────────────────────────────────────────────────────
    df = pd.read_csv('data/AIRS_clean.csv')
    print(f"\nLoaded: {df.shape[0]} rows x {df.shape[1]} columns")

    # Compute construct scores (mean of 2 items each)
    item_map = {
        'PE': ['PE1', 'PE2'], 'EE': ['EE1', 'EE2'], 'SI': ['SI1', 'SI2'],
        'FC': ['FC1', 'FC2'], 'HM': ['HM1', 'HM2'], 'PV': ['PV1', 'PV2'],
        'HB': ['HB1', 'HB2'], 'TR': ['TR1', 'TR2']
    }
    for construct, items in item_map.items():
        df[construct] = df[items].mean(axis=1)

    # BI outcome
    bi_items = ['BI1', 'BI2', 'BI3', 'BI4']
    df['BI'] = df[bi_items].mean(axis=1)

    print(f"\nConstruct scores computed (N = {len(df)}):")
    print(df[AIRS16_CONSTRUCTS + ['BI']].describe().round(3).to_string())

    # ── Correlation Matrix ────────────────────────────────────────────────
    corr = df[AIRS16_CONSTRUCTS + ['BI']].corr()
    print(f"\nCorrelation with BI:")
    for c in AIRS16_CONSTRUCTS:
        print(f"  {c}: r = {corr.loc[c, 'BI']:.3f}")

    # ── Prepare Clustering ────────────────────────────────────────────────
    X = df[AIRS16_CONSTRUCTS].dropna()
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)

    pop_means = df[AIRS16_CONSTRUCTS].mean()
    pop_stds = df[AIRS16_CONSTRUCTS].std()

    # ── Silhouette Analysis (k=2 to 8) ───────────────────────────────────
    print("\n" + "=" * 90)
    print("SILHOUETTE ANALYSIS")
    print("=" * 90)

    sil_scores = {}
    for k in range(2, 9):
        km = KMeans(n_clusters=k, n_init=10, random_state=RANDOM_STATE)
        labels = km.fit_predict(X_scaled)
        sil = silhouette_score(X_scaled, labels)
        sil_scores[k] = sil
        marker = ' ←' if k == SELECTED_K else ''
        print(f"  k={k}: silhouette = {sil:.3f}{marker}")

    # ── k=3 Final Clustering ──────────────────────────────────────────────
    print(f"\n" + "=" * 90)
    print(f"FINAL CLUSTERING (k={SELECTED_K})")
    print("=" * 90)

    km_final = KMeans(n_clusters=SELECTED_K, n_init=10, random_state=RANDOM_STATE)
    df['Cluster'] = km_final.fit_predict(X_scaled)
    final_silhouette = silhouette_score(X_scaled, df['Cluster'])
    print(f"  Silhouette score: {final_silhouette:.3f}")

    # Cluster sizes
    sizes = df['Cluster'].value_counts().sort_index()
    for c, n in sizes.items():
        print(f"  Cluster {c}: n = {n} ({n / len(df) * 100:.1f}%)")

    # Cluster profiles (z-scores)
    cluster_means_z = df.groupby('Cluster')[AIRS16_CONSTRUCTS].mean()
    for c in AIRS16_CONSTRUCTS:
        cluster_means_z[c] = (cluster_means_z[c] - pop_means[c]) / pop_stds[c]

    cluster_means_raw = df.groupby('Cluster')[AIRS16_CONSTRUCTS].mean()

    print("\n  Z-scored cluster profiles:")
    print(cluster_means_z.round(3).to_string())

    # Cluster naming based on profile patterns
    CLUSTER_NAMES = {}
    for cluster_id in range(SELECTED_K):
        profile = cluster_means_z.loc[cluster_id]
        mean_z = profile.mean()
        if mean_z > 0.3:
            CLUSTER_NAMES[cluster_id] = 'AI Enthusiasts'
        elif mean_z < -0.3:
            CLUSTER_NAMES[cluster_id] = 'AI Skeptics'
        else:
            CLUSTER_NAMES[cluster_id] = 'Moderate Users'

    df['Typology'] = df['Cluster'].map(CLUSTER_NAMES)

    print("\n  Cluster names:")
    for cid, name in CLUSTER_NAMES.items():
        n = sizes[cid]
        bi_mean = df[df['Cluster'] == cid]['BI'].mean()
        print(f"  {cid}: {name} (n={n}, BI mean={bi_mean:.2f})")

    # ── k=4 Comparison ────────────────────────────────────────────────────
    print(f"\n" + "=" * 90)
    print(f"k={K_ALT} ALTERNATIVE COMPARISON")
    print("=" * 90)

    km_k4 = KMeans(n_clusters=K_ALT, n_init=10, random_state=RANDOM_STATE)
    k4_labels = km_k4.fit_predict(X_scaled)
    sil_k4 = silhouette_score(X_scaled, k4_labels)
    k4_sizes = pd.Series(k4_labels).value_counts().sort_index()

    print(f"  k=4 silhouette: {sil_k4:.3f} (k=3: {final_silhouette:.3f})")
    for c, n in k4_sizes.items():
        print(f"  Cluster {c}: n = {n} ({n / len(df) * 100:.1f}%)")

    # k=4 z-scored profiles
    df_k4 = df.copy()
    df_k4['K4_Cluster'] = k4_labels
    k4_means = df_k4.groupby('K4_Cluster')[AIRS16_CONSTRUCTS].mean()
    k4_means_z = k4_means.copy()
    for c in AIRS16_CONSTRUCTS:
        k4_means_z[c] = (k4_means[c] - pop_means[c]) / pop_stds[c]

    print("\n  k=4 z-scored profiles:")
    print(k4_means_z.round(3).to_string())

    k4_defensible = sil_k4 >= 0.25 and k4_sizes.min() >= 50
    print(f"\n  k=4 defensible: {k4_defensible} (sil≥0.25={sil_k4 >= 0.25}, min_n≥50={k4_sizes.min() >= 50})")

    # ── Predictive Validation ─────────────────────────────────────────────
    print(f"\n" + "=" * 90)
    print("PREDICTIVE VALIDATION")
    print("=" * 90)

    # ANOVA: BI by cluster
    groups = [df[df['Cluster'] == c]['BI'].values for c in range(SELECTED_K)]
    f_stat, p_value = scipy_stats.f_oneway(*groups)
    # Eta squared
    grand_mean = df['BI'].mean()
    ss_between = sum(len(g) * (g.mean() - grand_mean)**2 for g in groups)
    ss_total = sum((df['BI'] - grand_mean)**2)
    eta_squared = ss_between / ss_total

    effect_interp = 'large' if eta_squared > 0.14 else ('medium' if eta_squared > 0.06 else 'small')
    print(f"  ANOVA: F = {f_stat:.2f}, p = {p_value:.6f}")
    print(f"  η² = {eta_squared:.3f} ({effect_interp})")

    # Regression: AIRS-16 → BI
    X_reg = df[AIRS16_CONSTRUCTS]
    y_reg = df['BI']
    reg_model = LinearRegression()
    reg_model.fit(X_reg, y_reg)
    r2 = reg_model.score(X_reg, y_reg)

    # Cross-validated R²
    cv_scores = cross_val_score(LinearRegression(), X_reg, y_reg, cv=5, scoring='r2')
    print(f"  Regression R² = {r2:.3f}")
    print(f"  5-fold CV R² = {cv_scores.mean():.3f} (±{cv_scores.std():.3f})")

    # Cluster-based R²
    df['BI_pred_cluster'] = df['Cluster'].map(df.groupby('Cluster')['BI'].mean())
    ss_res_cluster = sum((df['BI'] - df['BI_pred_cluster'])**2)
    r2_cluster = 1 - ss_res_cluster / ss_total
    print(f"  Cluster-based R² = {r2_cluster:.3f}")

    # ── Population Parameters ─────────────────────────────────────────────
    POP_PARAMS = {
        'means': pop_means.to_dict(),
        'stds': pop_stds.to_dict()
    }

    # ── Store Centroids ───────────────────────────────────────────────────
    CENTROIDS = {}
    for cluster_id in range(SELECTED_K):
        mask = df['Cluster'] == cluster_id
        CENTROIDS[CLUSTER_NAMES[cluster_id]] = {
            'cluster_id': cluster_id,
            'z_scores': cluster_means_z.loc[cluster_id].to_dict(),
            'raw_means': cluster_means_raw.loc[cluster_id].to_dict(),
            'expected_BI': float(df.loc[mask, 'BI'].mean()),
            'n': int(mask.sum()),
            'pct': float(mask.sum() / len(df) * 100)
        }

    # ── Classification Function Validation ────────────────────────────────
    def classify_airs16_typology(scores_dict):
        """Classify a respondent using stored centroids."""
        z_scores = {c: (scores_dict[c] - POP_PARAMS['means'][c]) / POP_PARAMS['stds'][c]
                    for c in AIRS16_CONSTRUCTS}
        distances = {}
        for name, data in CENTROIDS.items():
            dist = np.sqrt(sum((z_scores[c] - data['z_scores'][c])**2 for c in AIRS16_CONSTRUCTS))
            distances[name] = dist
        nearest = min(distances, key=distances.get)
        min_dist = distances[nearest]
        return nearest, 1 / (1 + min_dist)

    # Validate: reclassify all respondents
    correct = 0
    for idx, row in df.iterrows():
        scores = {c: row[c] for c in AIRS16_CONSTRUCTS}
        predicted, conf = classify_airs16_typology(scores)
        if predicted == row['Typology']:
            correct += 1
    agreement = correct / len(df)
    print(f"\n  Classification validation: {agreement * 100:.1f}% agreement")

    # ── Success Criteria ──────────────────────────────────────────────────
    print(f"\n" + "=" * 90)
    print("SUCCESS CRITERIA")
    print("=" * 90)
    print(f"  Silhouette > 0.25: {'PASS' if final_silhouette > 0.25 else 'FAIL'} ({final_silhouette:.3f})")
    print(f"  BI η² > 0.14: {'PASS' if eta_squared > 0.14 else 'FAIL'} ({eta_squared:.3f})")
    print(f"  Regression R² > 0.40: {'PASS' if r2 > 0.40 else 'FAIL'} ({r2:.3f})")

    # ── Export ────────────────────────────────────────────────────────────
    print(f"\n" + "=" * 90)
    print("EXPORTING RESULTS")
    print("=" * 90)

    results = {
        'metadata': {
            'notebook': '11_Practitioner_Typology',
            'date': pd.Timestamp.now().isoformat(),
            'n_samples': len(df),
            'constructs_used': AIRS16_CONSTRUCTS,
            'outcome_variable': OUTCOME
        },
        'clustering': {
            'n_clusters': SELECTED_K,
            'silhouette_score': float(final_silhouette),
            'cluster_sizes': {int(k): int(v) for k, v in sizes.items()}
        },
        'k4_alternative': {
            'n_clusters': K_ALT,
            'silhouette_score': float(sil_k4),
            'cluster_sizes': {int(k): int(v) for k, v in k4_sizes.items()},
            'defensible': bool(k4_defensible),
            'centroids_z': k4_means_z.round(3).to_dict()
        },
        'validation': {
            'bi_anova_f': float(f_stat),
            'bi_anova_p': float(p_value),
            'bi_eta_squared': float(eta_squared),
            'regression_r2': float(r2),
            'regression_cv_r2': float(cv_scores.mean()),
            'cluster_r2': float(r2_cluster)
        },
        'population_parameters': POP_PARAMS,
        'centroids': CENTROIDS,
        'cluster_names': CLUSTER_NAMES
    }

    results_path = Path('results/phase11_practitioner_typology_results.json')
    results_path.parent.mkdir(parents=True, exist_ok=True)
    with open(results_path, 'w') as f:
        json.dump(results, f, indent=2, default=str)
    print(f"✓ {results_path}")

    # ── Final Summary ─────────────────────────────────────────────────────
    print(f"\n" + "=" * 90)
    print("✓ PHASE 11: PRACTITIONER TYPOLOGY COMPLETE")
    print("=" * 90)
    print(f"  k=3 silhouette: {final_silhouette:.3f}")
    print(f"  k=4 silhouette: {sil_k4:.3f} (defensible: {k4_defensible})")
    for name, data in CENTROIDS.items():
        print(f"  {name}: n={data['n']} ({data['pct']:.1f}%), BI={data['expected_BI']:.2f}")


if __name__ == '__main__':
    main()
